src/monmove.cpp

```diff
@@ -101,16 +101,19 @@ bool monster::is_immune_field( const field_type_id fid ) const
 bool monster::can_move_to( const tripoint &p ) const
 {
     const bool can_climb = has_flag( MF_CLIMBS ) || has_flag( MF_FLIES );
-    if( g->m.impassable( p ) && !( can_climb && g->m.has_flag( "CLIMBABLE", p ) ) ) {
+    const bool non_diggable = g->m.has_flag( "ROAD", p ) ||
+                              g->m.has_flag( "WALL", p ) ||
+                              g->m.has_flag( "SUPPORTS_ROOF", p );
+    if( g->m.impassable( p ) &&
+        !( can_climb && g->m.has_flag( "CLIMBABLE", p ) ) &&
+        digging() && non_diggable ) {
```

Comment:
The check for digging and diggable should be in a separate if block. Currently the whole condition is false for all monsters that don't dig (`digging()` return `false`), which makes the whole condition false **independent** of the remaining checks.

For impassable terrain this becomes `true && ( ... ) && false && ...` and this is `false`.